Firestore/core/src/remote/bloom_filter.cc

```diff
@@ -0,0 +1,141 @@
+
+  CC_MD5_CTX context;
+  CC_MD5_Init(&context);
+  CC_MD5_Update(&context, key.data(), key.size());
+  CC_MD5_Final(md5_digest, &context);
+
+  uint64_t* hash128 = reinterpret_cast<uint64_t*>(md5_digest);
+  return BloomFilter::Hash{hash128[0], hash128[1]};
+}
+
+/**
+ * Calculate the ith hash value based on the hashed 64 bit unsigned integers,
+ * and calculate its corresponding bit index in the bitmap to be checked.
+ */
+int32_t BloomFilter::GetBitIndex(const BloomFilter::Hash& hash,
+                                 int32_t i,
+                                 int32_t bit_count) const {
+  uint64_t val = hash.h1 + i * hash.h2;
+  return val % bit_count;
+}
+
+/** Return whether the bit at the given index in the bitmap is set to 1. */
+bool BloomFilter::IsBitSet(const std::vector<uint8_t>& bitmap,
+                           int32_t index) const {
+  uint8_t byteAtIndex = bitmap[index / 8];
+  int offset = index % 8;
+  return (byteAtIndex & (0x01 << offset)) != 0;
+}
+
+BloomFilter::BloomFilter(std::vector<uint8_t> bitmap,
+                         int32_t padding,
+                         int32_t hash_count) {
+  HARD_ASSERT(padding >= 0 && padding < 8, "Invalid padding: %s",
+              std::to_string(padding));
+  HARD_ASSERT(hash_count >= 0, "Invalid hash count: %s",
+              std::to_string(hash_count));
+  // Only empty bloom filter can have 0 hash count.
+  HARD_ASSERT(bitmap.size() == 0 || hash_count != 0, "Invalid hash count:%s",
+              std::to_string(hash_count));
+  // Empty bloom filter should have 0 padding.
+  HARD_ASSERT(bitmap.size() != 0 || padding == 0,
+              "Expected padding of 0 when bitmap length is 0, but got %s",
+              std::to_string(padding));
+
+  bitmap_ = bitmap;
+  hash_count_ = hash_count;
+  bit_count_ = bitmap.size() * 8 - padding;
+}
+
+StatusOr<BloomFilter> BloomFilter::Create(std::vector<uint8_t> bitmap,
+                                          int32_t padding,
+                                          int32_t hash_count) {
+  if (padding < 0 || padding >= 8) {
+    return Status(firestore::Error::kErrorInvalidArgument,
+                  "Invalid padding: " + std::to_string(padding));
+  }
+  if (hash_count < 0) {
+    return Status(firestore::Error::kErrorInvalidArgument,
+                  "Invalid hash count: " + std::to_string(hash_count));
+  }
+  if (bitmap.size() > 0 && hash_count == 0) {
+    // Only empty bloom filter can have 0 hash count.
+    return Status(firestore::Error::kErrorInvalidArgument,
+                  "Invalid hash count: " + std::to_string(hash_count));
+  }
+  if (bitmap.size() == 0 && padding != 0) {
+    // Empty bloom filter should have 0 padding.
+    return Status(firestore::Error::kErrorInvalidArgument,
+                  "Expected padding of 0 when bitmap length is 0, but got " +
+                      std::to_string(padding));
+  }
+
+  return BloomFilter(bitmap, padding, hash_count);
```

Comment:
nit: Move from the `bitmap` here too: replace `bitmap` with `std::move(bitmap)`.